provisioning/service/src/Exceptions/DeviceProvisioningServiceException.cs

```diff
@@ -0,0 +1,120 @@
+        /// <param name="innerException">The inner exception</param>
+        /// <param name="isTransient">True if the error is transient and the operation should be retried.</param>
+        internal DeviceProvisioningServiceException(string message, Exception innerException, bool isTransient)
+            : base(message, innerException)
+        {
+            IsTransient = isTransient;
+        }
+
+        /// <summary>
+        /// Creates an instance of this class.
+        /// </summary>
+        /// <param name="message">The message.</param>
+        /// <param name="statusCode">The 3-digit HTTP status code returned by Device Provisioning Service.</param>
+        /// <param name="fields">The HTTP headers.</param>
+        internal DeviceProvisioningServiceException(string message, HttpStatusCode statusCode, IDictionary<string, string> fields)
+            : base(message)
+        {
+            IsTransient = DetermineIfTransient(statusCode);
+            StatusCode = statusCode;
+            Fields = fields;
+        }
+
+        /// <summary>
+        /// Creates an instance of this class.
+        /// </summary>
+        /// <param name="message">The message.</param>
+        /// <param name="statusCode">The 3-digit HTTP status code returned by Device Provisioning Service.</param>
+        /// <param name="errorCode">The specific 6-digit error code in the DPS response, if available.</param>
+        /// <param name="trackingId">Service reported tracking Id. Use this when reporting a service issue.</param>
+        /// <param name="fields">The HTTP headers.</param>
+        internal DeviceProvisioningServiceException(string message, HttpStatusCode statusCode, int errorCode, string trackingId, IDictionary<string, string> fields)
+            : base(message)
+        {
+            IsTransient = DetermineIfTransient(statusCode);
+            StatusCode = statusCode;
+            ErrorCode = errorCode;
+            TrackingId = trackingId;
+            Fields = fields;
+        }
+
+        /// <summary>
+        /// True if the error is transient.
+        /// </summary>
+        public bool IsTransient { get; private set; }
+
+        /// <summary>
+        /// Service reported tracking Id. Use this when reporting a service issue.
+        /// </summary>
+        public string TrackingId { get; private set; }
+
+        /// <summary>
+        /// The 3-digit HTTP status code returned by Device Provisioning Service.
+        /// </summary>
+        public HttpStatusCode StatusCode { get; private set; }
+
+        /// <summary>
+        /// The specific 6-digit error code in the DPS response, if available.
+        /// </summary>
+        public int ErrorCode { get; private set; }
+
+        /// <summary>
+        /// The HTTP headers.
+        /// </summary>
+        /// <remarks>
+        /// This is used by DPS E2E tests.
+        /// </remarks>
+        public IDictionary<string, string> Fields { get; private set; }
```

Comment:
Let's make sure this has a default instance, if not specified. We want to avoid accidental NRE errors by customers who don't think to check for null.